pkg/reconciler/managed/reconciler.go

```diff
@@ -581,6 +628,18 @@ func (r *Reconciler) Reconcile(_ context.Context, req reconcile.Request) (reconc
 		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 	}
 
+	// A resource would only be pending creation at this point if we failed
+	// to persist our annotations after the ExternalClient's Create method
+	// was called. If that is the case we may have lost a critical update to
+	// the external name and leaked a resource. The safest thing to do is to
+	// refuse to proceed.
+	if meta.GetExternalCreatePending(managed) != nil {
+		log.Debug(errCreatePending)
+		record.Event(managed, event.Warning(reasonCannotInitialize, errors.New(errCreatePending)))
+		managed.SetConditions(xpv1.ReconcileError(errors.New(errCreatePending)))
+		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
```

Comment:
Heh, I just kicked off a test and every managed resource immediately got stuck on this clause. I see that all managed resources have both a create pending and a successful create timestamp set at roughly the same time, for example:

```yaml
apiVersion: ec2.aws.crossplane.io/v1beta1
kind: InternetGateway                                                                    
metadata:           
  annotations:     
    crossplane.io/composition-resource-name: internetgateway
    crossplane.io/external-create-pending: "2021-08-31T07:29:04Z"
    crossplane.io/external-create-succeeded: "2021-08-31T07:29:04Z"
    crossplane.io/external-name: igw-08657141fbbe34faf
```

Adding the successful creation annotation should remove the creation pending annotation, so I guess the pending annotation is somehow being immediately re-added? I need to sign off for the evening, so I'll work through this tomorrow.

Reply:
I figured out the problem - I was using `meta.AddAnnotations(o, a)` rather than `o.SetAnnotations(a)`. I'm a little worried that using `SetAnnotations` in our tight retry loop might 'undo' annotations being added by another controller, so I've refactored how the annotations are expected to work in a separate commit.